src/inmanta/module.py

```diff
@@ -226,14 +232,110 @@ def get_file_for_version(self, repo: str, tag: str, file: str) -> str:
+        """
+        module_name: str = self._get_module_name(module_spec)
+        path: Optional[str] = self.path_for(module_name)
+        if path is not None:
+            return self.from_path(project, path)
+        elif install:
+            return self.install(project, module_spec)
+        else:
+            return None
+
+    @abstractmethod
+    def install(self, project: Optional["Project"], module_spec: List[Requirement]) -> Optional[TModule]:
+        """
+        Attempt to install a module given a module spec.
+
+        :param project: The project associated with the module, if any.
+        :param module_spec: The module specification including any constraints on its version.
+        """
+        raise NotImplementedError("Abstract method")
+
+    def path_for(self, name: str) -> Optional[str]:
+        raise NotImplementedError("Abstract method")
+
+    @classmethod
+    @abstractmethod
+    def from_path(cls, project: Optional["Project"], path: str) -> TModule:
         raise NotImplementedError("Abstract method")
 
+    def _get_module_name(self, module_spec: List[Requirement]) -> str:
+        module_names: Set[str] = {req.key for req in module_spec}
+        module_name: str = more_itertools.one(
+            module_names,
+            too_short=ValueError("module_spec should contain at least one requirement"),
+            too_long=ValueError("module_spec should contain requirements for exactly one module"),
+        )
+        return module_name
+
+
+class ModuleV2Source(ModuleSource["ModuleV2"]):
+    def __init__(self, urls: List[str]) -> None:
+        self.urls: List[str] = urls
+
+    def install(self, project: Optional["Project"], module_spec: List[Requirement]) -> Optional["ModuleV2"]:
+        module_name: str = self._get_module_name(module_spec)
+        requirements: List[Requirement] = [Requirement.parse(f"{ModuleV2.PKG_NAME_PREFIX}{str(req)}") for req in module_spec]
+        try:
+            env.ProcessEnv.install_from_indexes(requirements, self.urls)
+        except env.PackageNotFound:
+            return None
+        path: Optional[str] = self.path_for(module_name)
+        if path is None:
+            raise Exception(f"Inconsistent state: installed module {module_name} but it does not exist in the Python environment.")
+        return self.from_path(project, path)
+
     def path_for(self, name: str) -> Optional[str]:
-        # same class is used for search parh and remote repos, perhaps not optimal
+        if name.startswith(ModuleV2.PKG_NAME_PREFIX):
+            raise Exception("PythonRepo instances work with inmanta module names, not Python package names.")
+        package: str = f"inmanta_plugins.{name}"
+        init: Optional[str] = env.ProcessEnv.get_module_file(package)
+        if init is None:
+            return None
+        try:
+            return ModuleLike.get_first_directory_containing_file(os.path.dirname(init), ModuleV2.MODULE_FILE)
```

Comment:
this I don't understand

Reply:
Do you mean the whole method logic? It checks if the `inmanta_plugins.the_module` package can be found in the active venv. If so, it goes up the directory tree starting from the `__init__.py` file until it finds a `ModuleV2.MODULE_FILE`, indicating the module root. It would be cleaner to use `return ModuleV2.get_module_dir(os.path.dirname(init))` here.

Reply:
The walking up is a bit weird here. Either the file is exactly where it is expected or the module is corrupt?

Reply:
The thing is that it's different for editable vs non-editable installs, which is why I thought it easiest to determine the module root by its defining config file.

Reply:
I've updated this to be like Arnaud did it in #3164.